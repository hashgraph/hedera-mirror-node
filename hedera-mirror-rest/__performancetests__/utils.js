#!/usr/bin/env node
/*-
 *
 * Hedera Mirror Node
 *
 * Copyright (C) 2019 Hedera Hashgraph, LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */

const fs = require('fs');
const yaml = require('js-yaml');

// Load config
process.env.CONFIG_NAME = 'perfTestConfig';
process.env.CONFIG_PATH = __dirname;
const config = require('../config');

const mustLoadYaml = fileName => {
  try {
    console.log(`Loading yaml file ${fileName}`);
    return yaml.safeLoad(fs.readFileSync(fileName, 'utf8'));
  } catch (err) {
    console.log(`Failed to load yaml file ${fileName}: ${err}`);
    process.exit(1);
  }
};

module.exports = {
  config: config,
  mustLoadYaml: mustLoadYaml
};
