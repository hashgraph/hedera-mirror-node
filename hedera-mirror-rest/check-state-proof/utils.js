/*-
 * ‌
 * Hedera Mirror Node
 * ​
 * Copyright (C) 2019 - 2020 Hedera Hashgraph, LLC
 * ​
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * ‍
 */
'uses strict';

// external libraries
const fs = require('fs');
const fetch = require('node-fetch');
const AbortController = require('abort-controller');

const base64StringToBuffer = (base64String) => {
  return Buffer.from(base64String, 'base64');
};

const replaceSpecialCharsWithUnderScores = (stringToFormat) => {
  return stringToFormat.replace(/[.@\-]/g, '_');
};

const makeStateProofDir = (transactionId, stateProofJson) => {
  const newDirPath = replaceSpecialCharsWithUnderScores(transactionId);
  fs.mkdirSync(newDirPath, {recursive: true});
  fs.writeFileSync(`${newDirPath}/apiResponse.json`, JSON.stringify(stateProofJson));
  console.log(`Supporting files and API response for the state proof will be stored in the directory ${newDirPath}`);
  return newDirPath;
};

const storeFile = (data, file, ext) => {
  const newFilePath = `${replaceSpecialCharsWithUnderScores(file)}.${ext}`;
  fs.writeFileSync(`${newFilePath}`, data, (err) => {
    if (err) throw err;
  });
};

const getAPIResponse = async (url) => {
  const controller = new AbortController();
  const timeout = setTimeout(
    () => {
      controller.abort();
    },
    60 * 1000 // in ms
  );

  console.log(`Requesting state proof files from ${url}...`);
  return await fetch(url, {signal: controller.signal})
    .then(async (response) => {
      if (!response.ok) {
        throw Error(response.statusText);
      }
      return response.json();
    })
    .catch((error) => {
      throw `Error fetching ${url}: ${error}`;
    })
    .finally(() => {
      clearTimeout(timeout);
    });
};

const readJSONFile = (filePath) => {
  const rawData = fs.readFileSync(filePath);
  return JSON.parse(rawData);
};

module.exports = {
  base64StringToBuffer,
  getAPIResponse,
  makeStateProofDir,
  readJSONFile,
  storeFile,
  replaceSpecialCharsWithUnderScores,
};
