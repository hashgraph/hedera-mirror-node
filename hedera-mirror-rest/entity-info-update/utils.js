/*-
 * ‌
 * Hedera Mirror Node
 * ​
 * Copyright (C) 2019 - 2021 Hedera Hashgraph, LLC
 * ​
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * ‍
 */

'use strict';

// external libraries
const _ = require('lodash');
const fs = require('fs');
const log4js = require('log4js');
const math = require('mathjs');

// local
const config = require('./config');

const logger = log4js.getLogger();

const constructEntity = (index, headerRow, entityRow) => {
  const entityObj = {};
  const splitEntityRow = Array.from(entityRow.split(',')).filter((x) => x != null);

  for (let i = 0; i < headerRow.length; i++) {
    entityObj[headerRow[i].trim()] = splitEntityRow[i].trim();
  }

  return entityObj;
};

const readEntityCSVFileSync = () => {
  logger.info(`Parsing csv entity file ...`);
  const entities = [];

  const data = fs.readFileSync(config.filePath, 'utf-8');

  const fileContent = data.split('\n');
  const headers = Array.from(fileContent[0].split(',')).filter((x) => x != null);

  // ensure 1st column is entity num of expected format
  if (!_.eq(headers[0], 'entity')) {
    throw Error("CSV must have a header column with 1st column being 'entity'");
  }

  for (let i = 1; i < fileContent.length; i++) {
    if (!fileContent[i]) {
      // End of file
      break;
    }

    entities.push(constructEntity(i, headers, fileContent[i]));
  }

  logger.info(`${entities.length} entities were extracted from ${config.filePath}`);

  return entities;
};

/**
 * Converts seconds since epoch (seconds nnnnnnnnn format) to  nanoseconds
 * @param {String} Seconds since epoch (seconds.nnnnnnnnn format)
 * @return {String} ns Nanoseconds since epoch
 */
const secNsToNs = (secNs, ns) => {
  return math
    .add(math.multiply(math.bignumber(secNs.toString()), math.bignumber(1e9)), math.bignumber(ns.toString()))
    .toString();
};

module.exports = {
  readEntityCSVFileSync,
  secNsToNs,
};
