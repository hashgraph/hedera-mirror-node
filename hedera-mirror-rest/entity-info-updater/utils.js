/*-
 * ‌
 * Hedera Mirror Node
 * ​
 * Copyright (C) 2019 - 2021 Hedera Hashgraph, LLC
 * ​
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * ‍
 */

'use strict';

// external libraries

const proto = require('@hashgraph/proto');
const {KeyList, PublicKey} = require('@hashgraph/sdk');
const _ = require('lodash');
const fs = require('fs');
const log4js = require('log4js');
const path = require('path');

// local
const config = require('./config');

const logger = log4js.getLogger();

// used for timestamp range
const longMinValue = -9223372036854775808n;
const longMaxValue = 9223372036854775807n;

// used for entityId conversion
const realmOffset = 2n ** 32n; // realm is followed by 32 bits entity_num
const shardOffset = 2n ** 48n; // shard is followed by 16 bits realm and 32 bits entity_num

// sample location details
const entityCachePath = process.env.ENTITY_CACHE_PATH || `${__dirname}/sample`;
const dbCacheFileName = process.env.DB_ENTITY_CACHE_FILE || `dbEntityCache.json`;
const networkCacheFileName = process.env.NETWORK_ENTITY_CACHE_FILE || `networkEntityCache.json`;

const constructEntity = (index, headerRow, entityRow) => {
  const entityObj = {};
  const splitEntityRow = Array.from(entityRow.split(',')).filter((x) => x != null);

  for (let i = 0; i < headerRow.length; i++) {
    entityObj[headerRow[i].trim()] = splitEntityRow[i].trim();
  }

  return entityObj;
};

const readEntityCSVFileSync = () => {
  logger.info(`Parsing csv entity file ...`);
  const csvStart = process.hrtime();
  const entities = [];

  const data = fs.readFileSync(config.filePath, 'utf-8');

  const fileContent = data.split('\n');
  const headers = Array.from(fileContent[0].split(',')).filter((x) => x != null);

  // ensure first column is entity num of expected format
  if (!_.eq(headers[0], 'entity')) {
    throw Error("CSV must have a header column with first column being 'entity'");
  }

  for (let i = 1; i < fileContent.length; i++) {
    if (!fileContent[i]) {
      // End of file
      break;
    }

    entities.push(constructEntity(i, headers, fileContent[i]));
  }

  const elapsedTime = process.hrtime(csvStart);
  logger.info(
    `${entities.length} entities were extracted from ${config.filePath} in ${getElapsedTimeString(elapsedTime)}`
  );

  return entities;
};

/**
 * Converts timestamp of seconds since epoch (seconds nnnnnnnnn format) to  nanoseconds
 * @param {String} Seconds since epoch (seconds.nnnnnnnnn format)
 * @return {BigInt} ns Nanoseconds since epoch
 */
const timestampToNs = (sec, ns) => {
  let finalNs = 0;

  try {
    finalNs = BigInt(sec.toString()) * BigInt(1e9) + BigInt(ns.toString());
  } catch (e) {
    logger.warn(`Error parsing sec: ${JSON.stringify(sec)}, ns: ${JSON.stringify(ns)} to timestamp`);
    throw e;
  }

  // handle the equivalent long overflow case that the java importer accommodates
  if (finalNs < longMinValue) {
    return longMinValue;
  }

  if (finalNs > longMaxValue) {
    return longMaxValue;
  }

  return finalNs;
};

const keyListToProto = (keyList) => {
  const keys = [];

  for (const key of keyList) {
    keys.push(keyToProto(key));
  }

  return {
    keys,
  };
};

const keyToProto = (key) => {
  if (key instanceof PublicKey) {
    return {
      ed25519: key.toBytes(),
    };
  }

  if (key instanceof KeyList) {
    return {
      keyList: keyListToProto(key),
    };
  }

  throw Error('Unsupported key type');
};

const getProtoAndEd25519HexFromPublicKey = (key) => {
  const protoKey = {
    ed25519: key.toBytes(),
  };
  const ed25519Hex = key.toString();

  return {protoKey, ed25519Hex};
};

const getProtoAndEd25519HexFromKeyList = (key) => {
  const protoKey = {
    keyList: keyListToProto(key),
  };

  // only keyLists of length one have an applicable ed25519Hex
  let ed25519Hex = null;
  // some complex keys are nested, ensure top level key is the applicable ed25519Hex
  if (key._keys.length === 1 && key._keys[0]._keyData !== undefined) {
    ed25519Hex = key._keys[0].toString();
  }

  return {protoKey, ed25519Hex};
};

const getBufferAndEd25519HexFromKey = (key) => {
  const {protoKey, ed25519Hex} =
    key instanceof PublicKey ? getProtoAndEd25519HexFromPublicKey(key) : getProtoAndEd25519HexFromKeyList(key);
  let protoBuffer = null;
  protoBuffer = proto.Key.encode(protoKey, protoBuffer).finish();

  return {protoBuffer, ed25519Hex};
};

const getElapsedTimeString = (elapsedTime) => {
  return `${elapsedTime[0]} s ${elapsedTime[1] / 1000000} ms`;
};

/**
 * Converts sdk AccountId string to an encoded id corresponding to the Entity. Supports 'shard.realm.num' only
 *
 * @param {AccountId|null} entityIdStr
 * @return {string|null}
 */
const getEntityId = (entityIdStr) => {
  if (_.isNull(entityIdStr)) {
    return null;
  }

  // [shard,realm,num] - entityIdStr comes from network and will always conforming to the expected format
  const parts = entityIdStr.toString().split('.');

  return (BigInt(parts[2]) + BigInt(parts[1]) * realmOffset + BigInt(parts[0]) * shardOffset).toString();
};

const getNetworkEntityCache = () => {
  const pathString = path.join(entityCachePath, networkCacheFileName);
  return fs.existsSync(pathString) ? JSON.parse(fs.readFileSync(pathString, 'utf-8')) : {};
};

const getDbEntityCache = () => {
  const pathString = path.join(entityCachePath, dbCacheFileName);
  return fs.existsSync(pathString) ? JSON.parse(fs.readFileSync(pathString, 'utf-8')) : {};
};

module.exports = {
  getBufferAndEd25519HexFromKey,
  getElapsedTimeString,
  getDbEntityCache,
  getEntityId,
  getNetworkEntityCache,
  readEntityCSVFileSync,
  timestampToNs,
};
