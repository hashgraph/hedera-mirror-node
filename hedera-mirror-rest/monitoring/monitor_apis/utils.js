/*-
 * ‌
 * Hedera Mirror Node
 * ​
 * Copyright (C) 2019 - 2020 Hedera Hashgraph, LLC
 * ​
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * ‍
 */

'use strict';

const AbortController = require('abort-controller');
const httpErrors = require('http-errors');
const _ = require('lodash');
const fetch = require('node-fetch');
const querystring = require('querystring');
const config = require('./config');

const apiPrefix = '/api/v1';

/**
 * Converts shard.realm.accountNumber to accountNumber
 * @param {String} shard.realm.accountNumber
 * @return {Number} accountNumber
 */
const toAccNum = (accId) => Number(accId.split('.')[2]);

/**
 * Converts accountNumber to shard.realm.accountNumber string
 * @param {Number} accountNumber
 * @return {String} shard.realm.accountNumber
 */
const fromAccNum = (accNum) => `${config.shard}.0.${accNum}`;

/**
 * Create and return the url for a rest api call
 * If running on a local server http is employed over https
 * @param {Object} server
 * @param {String} path rest-api endpoint path
 * @param {Object} query key-value query params
 * @return {String} rest-api endpoint url
 */
const getUrl = (server, path, query = undefined) => {
  let endpoint = server;
  if (server.includes('localhost') || server.includes('127.0.0.1')) {
    endpoint = server.replace('https', 'http');
  }

  let url = `${endpoint}${apiPrefix}${path}`;
  if (query) {
    const qs = querystring.stringify(query);
    if (qs !== '') {
      url += `?${qs}`;
    }
  }

  return url;
};

/**
 * Make an http request to mirror-node api
 * Host info is prepended to if only path is provided
 * @param {*} url rest-api endpoint
 * @param {String} key JSON key of the object to return
 * @return {Object} JSON object representing api response or error
 */
const getAPIResponse = async (url, key = undefined) => {
  const controller = new AbortController();
  const timeout = setTimeout(
    () => {
      controller.abort();
    },
    60 * 1000 // in ms
  );

  try {
    const response = await fetch(url, {signal: controller.signal});
    if (!response.ok) {
      console.log(`Non success response for call to '${url}'`);
      return httpErrors(response.status, response.statusText);
    }

    const json = await response.json();
    return key ? json[key] : json;
  } catch (error) {
    const message = `Fetch error, url : ${url}, error : ${error}`;
    console.log(message);
    return Error(message);
  } finally {
    clearTimeout(timeout);
  }
};

class ServerTestResult {
  constructor() {
    this.result = {
      testResults: [],
      numPassedTests: 0,
      numFailedTests: 0,
      success: true,
      message: '',
      startTime: Date.now(),
      endTime: 0,
    };
  }

  addTestResult(testResult) {
    this.result.testResults.push(testResult);
    if (testResult.result === 'passed') {
      this.result.numPassedTests += 1;
    } else {
      this.result.numFailedTests += 1;
      this.result.success = false;
    }
  }

  finish() {
    this.result.endTime = Date.now();
  }
}

/**
 * Creates a function to run specific tests with the provided server address, classs result, and resource
 *
 * @param {String} server server address in the format of http://ip:port
 * @param {ServerTestResult} testClassResult test class result object
 * @param {String} resource name of the resource to test
 * @return {function(...[*]=)}
 */
const testRunner = (server, testClassResult, resource) => {
  return async (testFunc) => {
    const start = Date.now();
    const result = await testFunc(server);
    if (result.skipped) {
      return;
    }

    const testResult = {
      at: start,
      result: result.passed ? 'passed' : 'failed',
      url: result.url,
      message: result.passed ? result.message : '',
      failureMessages: !result.passed ? [result.message] : [],
      resource,
    };

    testClassResult.addTestResult(testResult);
  };
};

/**
 * Helper function to create a json object for failed test results
 * @param {String} title Title in the jest output
 * @param {String} msg Message in the jest output
 * @return {Object} Constructed failed result object
 */
const createFailedResultJson = (title, msg) => {
  const failedResultJson = new ServerTestResult().result;
  failedResultJson.numFailedTests = 1;
  failedResultJson.success = false;
  failedResultJson.message = 'Prerequisite tests failed';
  failedResultJson.testResults = [
    {
      at: (new Date().getTime() / 1000).toFixed(3),
      message: `${title}: ${msg}`,
      result: 'failed',
      assertionResults: [
        {
          ancestorTitles: title,
          failureMessages: [],
          fullName: `${title}: ${msg}`,
          location: null,
          status: 'failed',
          title: msg,
        },
      ],
    },
  ];

  return failedResultJson;
};

const checkAPIResponseError = (resp, option) => {
  const {expectHttpError, status} = option;
  const isRespError = resp instanceof Error;
  const isRespHttpError = resp instanceof httpErrors.HttpError;

  if (expectHttpError) {
    if (isRespHttpError && (!status || resp.status === status)) {
      return {passed: true};
    }

    const actual = isRespError ? JSON.stringify(resp) : '2xx';
    return {
      passed: false,
      message: `expect http error ${status || 'any'}, got ${actual}`,
    };
  }

  if (isRespError) {
    return {
      passed: false,
      message: resp.message,
    };
  }

  return {passed: true};
};

const checkRespObjDefined = (resp, option) => {
  const {message} = option;
  if (resp === undefined) {
    return {
      passed: false,
      message,
    };
  }
  return {passed: true};
};

const checkRespArrayLength = (elements, option) => {
  const {limit, message} = option;
  if (elements.length !== limit) {
    return {
      passed: false,
      message: message(elements, limit),
    };
  }
  return {passed: true};
};

const checkAccountNumber = (elements, option) => {
  const {accountNumber, message} = option;
  const element = Array.isArray(elements) ? elements[0] : elements;
  if (element.account !== fromAccNum(accountNumber)) {
    return {
      passed: false,
      message,
    };
  }
  return {passed: true};
};

const checkMandatoryParams = (elements, option) => {
  const element = Array.isArray(elements) ? elements[0] : elements;
  const {params, message} = option;
  for (let index = 0; index < params.length; index += 1) {
    if (!_.has(element, params[index])) {
      return {
        passed: false,
        message: `${message}: ${params[index]}`,
      };
    }
  }

  return {passed: true};
};

const checkRespDataFreshness = (resp, option) => {
  const {timestamp, threshold, message} = option;
  const ts = timestamp(Array.isArray(resp) ? resp[0] : resp);
  const secs = ts.split('.')[0];
  const currSecs = Math.floor(new Date().getTime() / 1000);
  const delta = currSecs - secs;
  if (delta > threshold) {
    return {
      passed: false,
      message: message(delta),
    };
  }

  return {passed: true};
};

const checkConsensusTimestampOrder = (elements, option) => {
  const {asc} = option;
  let previous = asc ? '0' : 'A';
  for (const element of elements) {
    const timestamp = element.consensus_timestamp;
    if (asc && timestamp <= previous) {
      return {passed: false, message: 'consensus timestamps are not in ascending order'};
    }
    if (!asc && timestamp >= previous) {
      return {passed: false, message: 'consensus timestamps are not in descending order'};
    }
    previous = timestamp;
  }

  return {passed: true};
};

/**
 * Checks resource freshness
 *
 * @param server the server address in the format of http://ip:port
 * @param path resource path
 * @param resource resource name
 * @param timestamp function to extract timestamp from response
 * @param jsonRespKey json response key to extract data from json response
 * @return {Promise<>}
 */
const checkResourceFreshness = async (server, path, resource, timestamp, jsonRespKey) => {
  const {freshnessThreshold} = config[resource];
  if (freshnessThreshold === 0) {
    return {skipped: true};
  }

  const url = getUrl(server, path, {limit: 1, order: 'desc'});
  const resp = await getAPIResponse(url, jsonRespKey);

  const checkRunner = new CheckRunner()
    .withCheckSpec(checkAPIResponseError)
    .withCheckSpec(checkRespObjDefined, {message: `${resource}: response object is undefined`});
  if (Array.isArray(resp)) {
    checkRunner.withCheckSpec(checkRespArrayLength, {
      limit: 1,
      message: (elements) => `${resource}: response data length of ${elements.length} was expected to be 1`,
    });
  }
  const result = checkRunner
    .withCheckSpec(checkRespDataFreshness, {
      timestamp,
      threshold: freshnessThreshold,
      message: (delta) => `${resource} was stale, ${delta} seconds old`,
    })
    .run(resp);
  if (!result.passed) {
    return {url, ...result};
  }

  return {
    url,
    passed: true,
    message: `Successfully retrieved ${resource} from with ${freshnessThreshold} seconds ago`,
  };
};

class CheckRunner {
  constructor() {
    this.checkSpecs = [];
  }

  withCheckSpec(check, option = {}) {
    this.checkSpecs.push({check, option});
    return this;
  }

  resetCheckSpec(check, option = {}) {
    this.checkSpecs.forEach((checkSpec) => {
      if (checkSpec.check === check) {
        checkSpec.option = option;
      }
    });
    return this;
  }

  run(data) {
    for (const checkSpec of this.checkSpecs) {
      const {check, option} = checkSpec;
      const result = check(data, option);
      if (!result.passed) {
        return result;
      }
    }

    return {passed: true};
  }
}

module.exports = {
  toAccNum,
  fromAccNum,
  getUrl,
  getAPIResponse,
  createFailedResultJson,
  testRunner,
  checkAPIResponseError,
  checkRespObjDefined,
  checkRespArrayLength,
  checkAccountNumber,
  checkMandatoryParams,
  checkConsensusTimestampOrder,
  checkResourceFreshness,
  CheckRunner,
  ServerTestResult,
};
