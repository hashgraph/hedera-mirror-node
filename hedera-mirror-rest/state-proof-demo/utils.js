/*-
 * ‌
 * Hedera Mirror Node
 * ​
 * Copyright (C) 2019 - 2020 Hedera Hashgraph, LLC
 * ​
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * ‍
 */
'uses strict';

// external libraries
const fs = require('fs');
const fetch = require('node-fetch');
const AbortController = require('abort-controller');

const base64StringToBuffer = (base64String) => {
  return Buffer.from(base64String, 'base64');
};

const replaceSpecialCharsWithUnderScores = (stringToFormat) => {
  return stringToFormat.replace(/\./g, '_').replace(/\@/g, '_').replace(/\-/g, '_');
};

const makeStateProofDir = (transactionId) => {
  fs.mkdir(transactionId, {recursive: true}, (err, path) => {
    if (err) {
      console.error(`Error encountered creating directory ${path}: ${err}`);
      throw err;
    }
  });

  fs.writeFile(
    `${transactionId}/notes.txt`,
    `Supporting files for the state proof of '${transactionId}' can be found in this directory`,
    (err) => {
      if (err) throw err;
    }
  );
};

const storeFile = (data, file, ext) => {
  const newFilePath = `${replaceSpecialCharsWithUnderScores(file)}.${ext}`;
  console.log(`Storing contents at ${newFilePath}`);
  fs.writeFile(`${newFilePath}`, data, (err) => {
    if (err) throw err;
    console.log(`File ${newFilePath} has been saved!`);
  });
};

const getAPIResponse = async (url) => {
  const controller = new AbortController();
  const timeout = setTimeout(
    () => {
      controller.abort();
    },
    60 * 1000 // in ms
  );

  console.log(`Requesting stateproof files from ${url}...`);
  return await fetch(url, {signal: controller.signal})
    .then(async (response) => {
      if (!response.ok) {
        throw Error(response.statusText);
      }
      return await response.json();
    })
    .catch((error) => {
      var message = `Fetch error, url : ${url}, error : ${error}`;
      console.log(message);
      throw message;
    })
    .finally(() => {
      clearTimeout(timeout);
    });
};

const readJSONFile = (filePath) => {
  let rawData = fs.readFileSync(filePath);
  return JSON.parse(rawData);
};

module.exports = {
  base64StringToBuffer,
  getAPIResponse,
  makeStateProofDir,
  readJSONFile,
  storeFile,
  replaceSpecialCharsWithUnderScores,
};
