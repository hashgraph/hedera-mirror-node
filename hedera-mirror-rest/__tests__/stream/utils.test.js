/*-
 * ‌
 * Hedera Mirror Node
 * ​
 * Copyright (C) 2019 - 2023 Hedera Hashgraph, LLC
 * ​
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * ‍
 */

import long from 'long';
import {proto} from '@hashgraph/proto';
import TransactionId from '../../transactionId';
import * as utils from '../../stream/utils';

describe('protoTransactionIdToTransactionId', () => {
  const accountID = {
    shardNum: 0,
    realmNum: 0,
    accountNum: 1010,
  };

  const testSpecs = [
    {
      transactionId: proto.TransactionID.create({
        accountID,
        transactionValidStart: {
          seconds: 193823,
          nanos: 0,
        },
      }),
      expected: TransactionId.fromString('0.0.1010-193823-0'),
    },
    {
      transactionId: proto.TransactionID.create({
        accountID,
        transactionValidStart: {
          seconds: 193823,
          nanos: 999999999,
        },
      }),
      expected: TransactionId.fromString('0.0.1010-193823-999999999'),
    },
    {
      transactionId: proto.TransactionID.create({
        accountID,
        transactionValidStart: {
          seconds: long.MAX_VALUE,
          nanos: 999999999,
        },
      }),
      expected: TransactionId.fromString(`0.0.1010-${long.MAX_VALUE.toString()}-999999999`),
    },
  ];

  testSpecs.forEach((testSpec) => {
    test(`expect output ${testSpec.expected}`, () => {
      expect(utils.protoTransactionIdToTransactionId(testSpec.transactionId)).toEqual(testSpec.expected);
    });
  });
});

describe('readLengthAndBytes', () => {
  const data = new Array(16).fill(0xab);
  const bufferWithChecksum = Buffer.from([0, 0, 0, 16, 0, 0, 0, 101 - 16].concat(data));
  const bufferNoChecksum = Buffer.from([0, 0, 0, 16].concat(data));

  const testSpecs = [
    {
      name: 'read from buffer with checksum',
      buffer: bufferWithChecksum,
      hasChecksum: true,
      minLength: 1,
      maxLength: data.length,
      expected: {
        length: bufferWithChecksum.length,
        bytes: Buffer.from(data),
      },
    },
    {
      name: 'read from buffer with checksum and exact length',
      buffer: bufferWithChecksum,
      hasChecksum: true,
      minLength: data.length,
      maxLength: data.length,
      expected: {
        length: bufferWithChecksum.length,
        bytes: Buffer.from(data),
      },
    },
    {
      name: 'read from buffer without checksum',
      buffer: bufferNoChecksum,
      hasChecksum: false,
      minLength: 1,
      maxLength: data.length,
      expected: {
        length: bufferNoChecksum.length,
        bytes: Buffer.from(data),
      },
    },
    {
      name: 'read from buffer with checksum error',
      buffer: bufferNoChecksum,
      hasChecksum: true,
      minLength: 1,
      maxLength: data.length,
    },
    {
      name: 'read from buffer with length over max',
      buffer: bufferWithChecksum,
      hasChecksum: true,
      minLength: 1,
      maxLength: data.length - 1,
    },
    {
      name: 'read from buffer with length smaller than min',
      buffer: bufferWithChecksum,
      hasChecksum: true,
      minLength: data.length + 1,
      maxLength: data.length + 2,
    },
    {
      name: 'read from buffer with length not match',
      buffer: bufferWithChecksum,
      hasChecksum: true,
      minLength: data.length + 1,
      maxLength: data.length + 1,
    },
  ];

  testSpecs.forEach((testSpec) => {
    const {name, buffer, minLength, maxLength, hasChecksum, expected} = testSpec;
    test(name, () => {
      if (expected) {
        const actual = utils.readLengthAndBytes(buffer, minLength, maxLength, hasChecksum);
        expect(actual).toEqual(expected);
      } else {
        expect(() =>
          utils.readLengthAndBytes(buffer, minLength, maxLength, hasChecksum)
        ).toThrowErrorMatchingSnapshot();
      }
    });
  });
});

describe('readNBytes', () => {
  const buffer = Buffer.from(new Array(16).fill(0xab));
  const testSpecs = [
    {
      name: 'read 1 byte from 16-byte buffer',
      buffer,
      length: 1,
      expected: buffer.slice(0, 1),
    },
    {
      name: 'read 16 byte from 16-byte buffer',
      buffer,
      length: 16,
      expected: buffer,
    },
    {
      name: 'read 0 byte from 16-byte buffer',
      buffer,
      length: 0,
      expected: buffer.slice(0, 0),
    },
    {
      name: 'read 17 byte from 16-byte buffer',
      buffer,
      length: 17,
    },
    {
      name: 'read -1 byte from 16-byte buffer',
      buffer,
      length: -1,
    },
  ];

  testSpecs.forEach((testSpec) => {
    test(testSpec.name, () => {
      if (testSpec.expected) {
        expect(utils.readNBytes(testSpec.buffer, testSpec.length)).toEqual(testSpec.expected);
      } else {
        expect(() => utils.readNBytes(testSpec.buffer, testSpec.length)).toThrowErrorMatchingSnapshot();
      }
    });
  });
});
