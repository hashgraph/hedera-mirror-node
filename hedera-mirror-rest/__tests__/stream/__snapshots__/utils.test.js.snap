// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`readLengthAndBytes read from buffer with checksum error 1`] = `"Error reading length and bytes, expect checksum -1414812757 to be 85"`;

exports[`readLengthAndBytes read from buffer with length not match 1`] = `"Error reading length and bytes, expect length 17 got 16"`;

exports[`readLengthAndBytes read from buffer with length over max 1`] = `"Error reading length and bytes, expect length 16 within [1, 15]"`;

exports[`readLengthAndBytes read from buffer with length smaller than min 1`] = `"Error reading length and bytes, expect length 16 within [17, 18]"`;

exports[`readNBytes read -1 byte from 16-byte buffer 1`] = `"Error reading byte array, expect -1-byte data got 16-byte"`;

exports[`readNBytes read 17 byte from 16-byte buffer 1`] = `"Error reading byte array, expect 17-byte data got 16-byte"`;
